helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -146,6 +148,16 @@ private void generateMessage(final Resource resource, final BaseControllerDataPr
         }
       }
 
+      // Look through the current state map and add DROPPED message if the instance is not in the
+      // resourceStateMap. This instance may not have had been dropped by the rebalance strategy.
+      // This check is required to ensure that the instances removed from the ideal state stateMap
+      // are properly dropped.
+      for (String instance : currentStateMap.keySet()) {
+        if (!instanceStateMap.containsKey(instance)) {
+          instanceStateMap.put(instance, HelixDefinedState.DROPPED.name());
```

Comment:
Make sure this is just temp usage. Not persist back to ZK.

Reply:
This map isn't persisted back, it is copied and temporarily used:
```
Map<String, String> instanceStateMap =
          new HashMap<>(resourcesStateMap.getInstanceStateMap(resourceName, partition));
```